clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/names/GoogleLabeledResourceNamer.java

```diff
@@ -0,0 +1,108 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.google.names;
+
+import com.netflix.frigga.Names;
+import com.netflix.spinnaker.clouddriver.google.model.GoogleLabeledResource;
+import com.netflix.spinnaker.clouddriver.names.NamingStrategy;
+import com.netflix.spinnaker.moniker.Moniker;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+import java.util.function.Consumer;
+
+@Component
+public class GoogleLabeledResourceNamer implements NamingStrategy<GoogleLabeledResource> {
+  static String GCE_MONIKER_PREFIX = "spinnaker-moniker-";
+  static String APP = GCE_MONIKER_PREFIX + "application";
+  static String CLUSTER = GCE_MONIKER_PREFIX + "cluster";
+  static String DETAIL = GCE_MONIKER_PREFIX + "detail";
+  static String STACK = GCE_MONIKER_PREFIX + "stack";
+  static String SEQUENCE = GCE_MONIKER_PREFIX + "sequence";
+
+  @Override
+  public String getName() {
+    return "gceAnnotations";
+  }
+
+  public void applyMoniker(GoogleLabeledResource labeledResource, Moniker moniker) {
+    Map<String, String> templateLabels = labeledResource.getLabels();
+    setIfPresent(value -> templateLabels.putIfAbsent(APP, value.toLowerCase()), moniker.getApp());
+    setIfPresent(value -> templateLabels.putIfAbsent(CLUSTER, value.toLowerCase()), moniker.getCluster());
+    setIfPresent(value -> templateLabels.putIfAbsent(DETAIL, value.toLowerCase()), moniker.getDetail());
+    setIfPresent(value -> templateLabels.putIfAbsent(STACK, value.toLowerCase()), moniker.getStack());
+    setIfPresent(value -> templateLabels.put(SEQUENCE, value),
+      moniker.getSequence() != null ? moniker.getSequence().toString() : null); // Always overwrite sequence
+  }
+
+  @Override
+  public Moniker deriveMoniker(GoogleLabeledResource labeledResource) {
+    String name = labeledResource.getName();
+    Names parsed = Names.parseName(name);
+
+    Moniker moniker = Moniker.builder()
+        .app(parsed.getApp())
+        .cluster(parsed.getCluster())
+        .detail(parsed.getDetail())
+        .stack(parsed.getStack())
+        .sequence(parsed.getSequence())
+        .build();
+
+    Map<String, String> labels = labeledResource.getLabels();
+    if (moniker.getApp() != null && labels != null) {
+      setIfPresent(moniker::setApp, labels.get(APP));
+      String cluster = labels.get(CLUSTER);
+      String stack = labels.get(STACK);
+      String detail = labels.get(DETAIL);
+      String sequence = labels.get(SEQUENCE);
+      if (cluster == null && (detail != null || stack != null)) {
+        // If detail or stack is set and not cluster, we generate the cluster name using frigga convention (app-stack-detail)
+        cluster = getClusterName(moniker.getApp(), stack, detail);
```

Comment:
This is to make sure that `stack` and `detail` is useful. As you probably know, the _only_ field Deck cares about when deciding which cluster to put a resource in, is `cluster`. If it is set, it will be used (ignoring `stack` and `detail`). Because we're always generating the moniker using Frigga first, cluster will always be set. Therefore, if `stack` or `detail` is set using labels, but not `cluster`, we generate a new cluster name (Frigga style) using the `stack` and `detail`.